packages/context/src/resolver.ts

```diff
@@ -134,6 +128,13 @@ function resolve<T>(
         return injection.resolve(ctx, injection, s);
       } else {
         // Default to resolve the value from the context by binding key
+        if (injection.metadata && injection.metadata.optional) {
+          // If the `optional` flag is set for the injection, the resolution
+          // will return `undefined` instead of throwing an error
+          if (!ctx.isBound(injection.bindingKey)) {
```

Comment:
I am worried that this check is duplicating work done in `ctx.getValueOrPromise`. I am proposing to rework `ctx.getValueOrPromise` to accept `options` argument, so that we can pass the `optional` flag and tell `ctx.getValueOrPromise` to return `undefined` instead of throwing an exception.

Ideally, the optional `session` argument should be moved to this new `options` argument too.

I envision the following usage in this method:

```ts
return ctx.getValueOrPromise(injection.bindingKey, {
  session: s,
  optional: injection.metadata.optional,
  // perhaps "ignoreUnknownKey" may be better name than "optional"?
});
```

This will allow us to add support for optional bindings to other existing APIs, for example `ctx.get` and `ctx.getAsync`.

```ts
const maybeValue = await app.get('key-not-defined', {optional: true});
```